File: android/app/src/main/java/com/francescooddo/buddygrammar/core/adaptive/TapWordAcceptancePolicy.kt
package com.francescooddo.buddygrammar.core.adaptive

/** Shared production thresholds for tap suggestions and automatic correction. */
enum class TapWordAcceptancePolicy(
    val minimumConfidence: Double,
    val minimumMargin: Double,
) {
    SUGGESTION(minimumConfidence = 0.38, minimumMargin = 0.08),
    AUTOMATIC(minimumConfidence = 0.50, minimumMargin = 0.18),
    ;

    fun acceptedCandidate(result: TapWordDecodingResult): TapWordCandidate? {
        val best = result.candidates.firstOrNull() ?: return null
        return best.takeIf {
            it.confidence >= minimumConfidence &&
                result.margin >= minimumMargin &&
                !it.word.equals(result.literalWord, ignoreCase = true)
        }
    }

    /**
     * Selects a production replacement through the same suppression seam as
     * spelling suggestions. This prevents tap-lattice candidates from
     * bypassing a user's explicit "Never suggest" decision.
     */
    fun acceptedReplacement(
        result: TapWordDecodingResult,
        visibleWord: String,
        isSuppressed: (typed: String, suggestion: String) -> Boolean,
    ): String? {
        val replacement = acceptedCandidate(result)?.word ?: return null
        return replacement.takeUnless {
            it.equals(visibleWord, ignoreCase = true) ||
                isSuppressed(visibleWord, it)
        }
    }
}

File: BuddyGrammarKit/Sources/BuddyGrammarKit/TapWordAcceptancePolicy.swift
import Foundation

/// Shared production thresholds for presenting versus automatically applying
/// a tap-lattice replacement. Keeping this policy in the core decoder prevents
/// iOS and Android controller call sites from drifting.
public enum TapWordAcceptancePolicy: String, Sendable {
    case suggestion
    case automatic

    public var minimumConfidence: Double {
        switch self {
        case .suggestion: 0.38
        case .automatic: 0.50
        }
    }

    public var minimumMargin: Double {
        switch self {
        case .suggestion: 0.08
        case .automatic: 0.18
        }
    }

    public func acceptedCandidate(
        from result: TapWordDecodingResult
    ) -> TapWordCandidate? {
        guard let best = result.candidates.first,
              best.confidence >= minimumConfidence,
              result.margin >= minimumMargin,
              best.word.caseInsensitiveCompare(result.literalWord) != .orderedSame else {
            return nil
        }
        return best
    }
}
